plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -520,11 +524,153 @@ int add_event(struct flb_cloudwatch *ctx, struct cw_flush *buf,
+    || strncmp(an_item->key.via.str.ptr, "Swap.total", 10) == 0 || strncmp(an_item->key.via.str.ptr, "Swap.used", 9) == 0 || strncmp(an_item->key.via.str.ptr, "Swap.free", 9) == 0)
+    {
+        return 1;
+    }
+
+    return 0;
+}
+
+struct msgpack_object pack_emf_payload(struct flb_cloudwatch *ctx, struct mk_list *flb_ir_metrics, const char *input_plugin, struct flb_time tms)
+{
+    struct mk_list *metric_temp;
+    struct mk_list *metric_head;
+
+    /* msgpack::sbuffer is a simple buffer implementation. */
+    msgpack_sbuffer sbuf_emf;
+    msgpack_sbuffer_init(&sbuf_emf);
+
+    /* Serialize values into the buffer using msgpack_sbuffer_write callback function. */
+    msgpack_packer packer_emf;
+    msgpack_packer_init(&packer_emf, &sbuf_emf, msgpack_sbuffer_write);
+    msgpack_pack_map(&packer_emf, (mk_list_size(flb_ir_metrics)) + 1);
+
+    //Pack the _aws map
+    msgpack_pack_str(&packer_emf, 4);
+    msgpack_pack_str_body(&packer_emf, "_aws", 4);
+
+    msgpack_pack_map(&packer_emf, 2);
+
+    msgpack_pack_str(&packer_emf, 9);
+    msgpack_pack_str_body(&packer_emf, "Timestamp", 9);
+    msgpack_pack_long_long(&packer_emf, tms.tm.tv_sec * 1000L);
+
+    msgpack_pack_str(&packer_emf, 17);
+    msgpack_pack_str_body(&packer_emf, "CloudWatchMetrics", 17);
+    msgpack_pack_array(&packer_emf, 1);
+
+    msgpack_pack_map(&packer_emf, 3);
+
+    msgpack_pack_str(&packer_emf, 9);
+    msgpack_pack_str_body(&packer_emf, "Namespace", 9);
+
+    if (ctx->metric_namespace)
+    {
+        msgpack_pack_str(&packer_emf, flb_sds_len(ctx->metric_namespace));
+        msgpack_pack_str_body(&packer_emf, ctx->metric_namespace, flb_sds_len(ctx->metric_namespace));
+    }
+    else
+    {
+        msgpack_pack_str(&packer_emf, 18);
+        msgpack_pack_str_body(&packer_emf, "fluent-bit-metrics", 18);
+    }
+
+    msgpack_pack_str(&packer_emf, 10);
+    msgpack_pack_str_body(&packer_emf, "Dimensions", 10);
+
+    struct mk_list *head;
+    struct flb_split_entry *dimension_list;
+    if (ctx->metric_dimensions)
+    {
+        msgpack_pack_array(&packer_emf, mk_list_size(ctx->metric_dimensions));
+        mk_list_foreach(head, ctx->metric_dimensions)
+        {
+            dimension_list = mk_list_entry(head, struct flb_split_entry, _head);
+            struct mk_list *csv_values = flb_utils_split(dimension_list->value, ',', 256);
+            msgpack_pack_array(&packer_emf, mk_list_size(csv_values));
+
+            struct flb_split_entry *entry;
+            struct mk_list *inner_head;
+            mk_list_foreach(inner_head, csv_values)
+            {
+                entry = mk_list_entry(inner_head, struct flb_split_entry, _head);
+                msgpack_pack_str(&packer_emf, strlen(entry->value));
+                msgpack_pack_str_body(&packer_emf, entry->value, strlen(entry->value));
+            }
+        }
```

Comment:
if I am not wrong csv_values is going out of the scope leading to a memory leak

Reply:
I think you're right, @hossain-rayhan `flb_utils_split` allocates memory in each iteration of the loop...

Reply:
Memory freed. I think it should help.